### ISSUE TITLE
Crash during initial analysis with a custom architecture

### ISSUE THREAD

**Issue by whitequark** [Vector35/binaryninja-api#5533]
**Version and Platform (required):**
- Binary Ninja Version: 4.1.5442-dev Personal (66290799)
- OS: Windows
- OS Version: 10.0.19045 Build 19045
- CPU Architecture: x64

**Bug Description:**
During initial analysis of a firmware using a custom architecture, Binary Ninja crashes. This should not happen even if the architecture is at fault because it makes it impossible to determine the cause of the crash and address it.

**Steps To Reproduce:**
1. Install the [Avnera plugin](https://github.com/whitequark/binja-avnera) at commit 65bdae755de4e92f9549ce95e2db3b413a856976.
2. Download and extract [firmware-avnera-0xbb5c.zip](https://github.com/user-attachments/files/15755294/firmware-avnera-0xbb5c.zip)
3. Open the binary `firmware-avnera-0xbb5c.bin`
4. Set "Platform" to "avnera"
5. Set "Image Base Address" to "0xbb5c"
6. Make sure "Analysis Mode" is set to "full"
7. Click "Open"
8. Wait. There are some messages related to failed disassembly and such. The base address is actually incorrect and this is expected.
9. Binary Ninja suddenly crashes, seemingly always at the same progress of analysis.

**Expected Behavior:**
No crash. Some kind of message in the log indicates a problem with the plugin.

**Screenshots/Video Recording:**
I could record a video but it doesn't seem helpful.

**Binary:**
As above.

**Additional Information:**
The crash is reliably reproduced. I think I've also got the same (or a similar crash) on Linux so it doesn't seem platform-dependent.

I don't currently have access to my Linux machine and I don't debug native Windows applications a lot so there's no backtrace, but since the reproducer has been so reliable I feel like this is enough. I can get a backtrace if there's issues reproducing it.

Also, I tried catching all exceptions in the Python architecture callbacks (the commit I linked should be doing that), but that didn't help with avoiding crashes, so I think it's probably related to an uncaught API misuse?

**Comment by xusheng6** [comment]
This is a nullptr defef inside of our core. I have tracked down the crash site. I am not super familiar with the relevant code so I will ask the team to look at it

**Comment by whitequark** [comment]
Fantastic, thank you for such a very quick response!

**Comment by xusheng6** [comment]
Also just for your convenience -- your arch code should be able to create and set the avnera platform automatically, so that you do not have to repeat the Platform selection every time you load the binary

**Comment by whitequark** [comment]
Can I do that if the binary I'm working with doesn't have a useful header? (I could maybe pattern match on the bit of crt0 and the checksum but it doesn't feel reliable to me...)

**Comment by xusheng6** [comment]
> Can I do that if the binary I'm working with doesn't have a useful header? (I could maybe pattern match on the bit of crt0 and the checksum but it doesn't feel reliable to me...)

oh sorry I did not notice that. Then I am not sure whether it can be done in such a way

**Comment by xusheng6** [comment]
The crash has been fixed. 

On the other side, the lifter probably forgets to call MarkLabel() as required. I do not have enough time to checkout your code, nor do we have proper documentation on it (what a pity). However, you can look at our x86 architecture code and see how we do it: https://github.com/Vector35/binaryninja-api/blob/fd45eb84c513dc514ee8eb576ebb6a2816716987/arch/x86/il.cpp#L372-L376

**Comment by whitequark** [comment]
> However, you can look at our x86 architecture code and see how we do it:

I have used this pattern in another lifter. However, the problem with it is that it makes LLIL a lot less readable by littering it with jumps and extra edges. I found this a significant obstacle while debugging (and I had to look at LLIL a lot because this is a black box architecture).

Here are the only two culprits I can think of:

https://github.com/whitequark/binja-avnera/blob/6798c83d5a713fed1e21816d6a2b9c03eb8bdadd/mc/instr.py#L649-L664

https://github.com/whitequark/binja-avnera/blob/6798c83d5a713fed1e21816d6a2b9c03eb8bdadd/mc/instr.py#L686-L690

Does any of these patterns strike you as obviously broken? I thought I had avoided the (undocumented) pitfalls around labels but I suppose not.

**Comment by whitequark** [comment]
Also, thanks for fixing the crash so rapidly!

**Comment by xusheng6** [comment]
> > However, you can look at our x86 architecture code and see how we do it:
> 
> I have used this pattern in another lifter. However, the problem with it is that it makes LLIL a lot less readable by littering it with jumps and extra edges. I found this a significant obstacle while debugging (and I had to look at LLIL a lot because this is a black box architecture).
> 
> Here are the only two culprits I can think of:
> 
> https://github.com/whitequark/binja-avnera/blob/6798c83d5a713fed1e21816d6a2b9c03eb8bdadd/mc/instr.py#L649-L664
> 
> https://github.com/whitequark/binja-avnera/blob/6798c83d5a713fed1e21816d6a2b9c03eb8bdadd/mc/instr.py#L686-L690
> 
> Does any of these patterns strike you as obviously broken? I thought I had avoided the (undocumented) pitfalls around labels but I suppose not.

Yeah, unfortunately our vision is people write lifter to lift the code to LLIL, and they will then look at the HLIL to do the task. So your use case is new to us.

That said, if you can provide more details on the differences of the generated LLIL (maybe a screenshot of the code), I can see if you are doing it correctly. I am unable to determine it simply by reading the code

**Comment by whitequark** [comment]
> Yeah, unfortunately our vision is people write lifter to lift the code to LLIL, and they will then look at the HLIL to do the task. So your use case is new to us.

What happened with this architecture is that there were multiple issues where I wasn't lifting correctly, but this happened because I did not disassemble correctly. For example there was a carry/borrow mixup, there were multiple flag assignment issues, initially I implemented all of the jumps with the inverted condition.

Because there is no ISA manual or encoding diagrams or... anything, I had to infer what is actually meant by staring at the only two or three blobs I had really hard. And this meant looking at all of the ILs to find out where I am screwing things up. Mainly HLIL (because the ISA is not dense and rather awkward) and LLIL (because if you screw up lifting HLIL frequently becomes incomprehensible).

> That said, if you can provide more details on the differences of the generated LLIL (maybe a screenshot of the code), I can see if you are doing it correctly. I am unable to determine it simply by reading the code

The idea is that there are two options for conditional jumps. The first one is the pattern you've shown exactly as is. The second is for the case where (a) there is a label already at the taken and not taken addresses returned by `il.get_label_for_address` or (b) a label can be added to the current function by calling `il.add_label_for_address`. In this case I jump directly to the two labels instead of adding new ones to the function.

If I do it this way, I avoid the case where the LLIL looks approximately like this:

1\. some op
2\. some op
3\. if expr then 50 else 49
4\. some op
5\. jump 7
6\. some op
7\. ...
49\. jump 4
50\. jump 7

I think you can see why this LLIL is difficult to read while debugging.

Also, I would use the linear view a lot because I don't have any binaries where I would have any symbols or debug information or constants or defines or register addresses or *anything* to contextualize it, so the disassembly stream is literally the only thing I have. The graph view breaks the correspondence to disassembly while the linear view keeps it roughly 1:1, and this hack makes the correspondence a lot closer by eliding unnecessary jumps.

I hope this all makes sense.

**Comment by whitequark** [comment]
TL;DR can you imagine any issues using `il.add_label_for_address` while constructing LLIL for jump targets that are known to be within the current function (that is, relative jumps with pc-relative offset)?
